[Embedded] Do not produce `cannot_specialize_class` for live issues

bnbarham opened the pull request:
SourceKit explicitly disables WMO, silence the diagnostic in this case (but leave it enabled for explicit non-WMO builds otherwise).

bnbarham commented:
@swift-ci please test

kubamracek commented:
Makes sense to me. @eeckstein ?

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/PassManager/Options.swift:
```diff
@@ -68,4 +68,8 @@ struct Options {
     }
   }
+
+  var enablePerformanceDiagnostics: Bool {
```
This is a bad name. Originally, `SILOptions::EnablePerformanceDiagnostics` was only used to disable the PerformanceDiagnosticsPass pass. But in embedded swift, such diagnostics are not "performance" related at all.
Can we just call it `isInSourceKit` or something like that?
And add a comment that in SourceKit we can't emit certain diagnostics which rely on wmo, which SourceKit doesn't support.

Also, as this is not a command line option, we should move this property directly into `Context` (at least on the swift side).

bnbarham replied:
How would you feel about `skipWMORequiredDiagnostics` or `enableWMORequestDiagnostics` or something similar? I'd rather make this more generic than sourcekit specifically, even if it is the only place we do this today.

eeckstein replied:
Sounds reasonable. Please add a comment that this is (currently) the case in SourceKit

bnbarham commented:
@swift-ci please test

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/PassManager/ModulePassContext.swift:
```diff
@@ -26,6 +26,10 @@ struct ModulePassContext : Context, CustomStringConvertible {
     return String(taking: _bridged.getModuleDescription())
   }
 
+  var enableWMORequiredDiagnostics: Bool {
```
Can you put this directly into Context (into `extension Context`)? We might need this in function passes, too.

eeckstein commented on include/swift/AST/SILOptions.h:
```diff
@@ -153,9 +153,11 @@ class SILOptions {
-  /// Enables SIL-level performance diagnostics (for @noLocks, @noAllocation
-  /// annotations and for Embedded Swift).
-  bool EnablePerformanceDiagnostics = true;
+  /// Enable diagnostics requiring WMO (for @noLocks, @noAllocation
```
Please add this comment to `Context.enableWMORequiredDiagnostics`, too.

bnbarham committed:
[Embedded] Do not produce `cannot_specialize_class` for live issues

SourceKit explicitly disables WMO, silence the diagnostic in this case
(but leave it enabled for explicit non-WMO builds otherwise).

bnbarham commented:
@swift-ci please test

eeckstein reviewed (approved):
thanks, lgtm!
